metrics.go

```diff
@@ -0,0 +1,416 @@
+	return "c"
+}
+
+func (c CounterMetric) SerializeValue() string {
+	return fmt.Sprintf(":%v", c.value)
+}
+
+func NewCounterMetric(key string, unit MetricUnit, tags map[string]string, timestamp int, value float64) CounterMetric {
+	am := abstractMetric{
+		key,
+		unit,
+		tags,
+		timestamp,
+	}
+
+	return CounterMetric{
+		value,
+		am,
+	}
+}
+
+// Distribution Metric.
+type DistributionMetric struct {
+	values []float64
+	abstractMetric
+}
+
+func (d *DistributionMetric) Add(value float64) {
+	d.values = append(d.values, value)
+}
+
+func (d DistributionMetric) GetType() string {
+	return "d"
+}
+
+func (d DistributionMetric) SerializeValue() string {
+	var sb strings.Builder
+	for _, el := range d.values {
+		sb.WriteString(fmt.Sprintf(":%v", el))
+	}
+	return sb.String()
+}
+
+func NewDistributionMetric(key string, unit MetricUnit, tags map[string]string, timestamp int, value float64) DistributionMetric {
+	am := abstractMetric{
+		key,
+		unit,
+		tags,
+		timestamp,
+	}
+
+	return DistributionMetric{
+		[]float64{value},
+		am,
+	}
+}
+
+// Gauge Metric.
+type GaugeMetric struct {
+	last  float64
+	min   float64
+	max   float64
+	sum   float64
+	count float64
+	abstractMetric
+}
+
+func (g *GaugeMetric) Add(value float64) {
+	g.last = value
+	g.min = math.Min(g.min, value)
+	g.max = math.Max(g.max, value)
+	g.sum += value
+	g.count++
+}
+
+func (g GaugeMetric) GetType() string {
+	return "g"
+}
+
+func (g GaugeMetric) SerializeValue() string {
+	return fmt.Sprintf(":%v:%v:%v:%v:%v", g.last, g.min, g.max, g.sum, g.count)
+}
+
+func NewGaugeMetric(key string, unit MetricUnit, tags map[string]string, timestamp int, value float64) GaugeMetric {
+	am := abstractMetric{
+		key,
+		unit,
+		tags,
+		timestamp,
+	}
+
+	return GaugeMetric{
+		value, // last
+		value, // min
+		value, // max
+		value, // sum
+		value, // count
+		am,
+	}
+}
+
+// Set Metric.
+type SetMetric[T NumberOrString] struct {
+	values map[T]void
+	abstractMetric
+}
+
+func (s *SetMetric[T]) Add(value T) {
+	s.values[value] = member
+}
+
+func (s SetMetric[T]) GetType() string {
+	return "s"
+}
+
+func (s SetMetric[T]) SerializeValue() string {
+	_hash := func(s string) uint32 {
+		return crc32.ChecksumIEEE([]byte(s))
+	}
+
+	values := make([]T, 0, len(s.values))
+	for k := range s.values {
+		values = append(values, k)
+	}
+	sortSlice(values)
+
+	var sb strings.Builder
+	for _, el := range values {
+		switch any(el).(type) {
+		case int:
+			sb.WriteString(fmt.Sprintf(":%v", el))
+		case string:
+			s := fmt.Sprintf("%v", el)
+			sb.WriteString(fmt.Sprintf(":%d", _hash(s)))
+		}
+	}
+
+	return sb.String()
+}
+
+func NewSetMetric[T NumberOrString](key string, unit MetricUnit, tags map[string]string, timestamp int, value T) SetMetric[T] {
+	am := abstractMetric{
+		key,
+		unit,
+		tags,
+		timestamp,
+	}
+
+	return SetMetric[T]{
+		map[T]void{
+			value: member,
+		},
+		am,
+	}
+}
+
+func sanitizeKey(s string) string {
+	re := regexp.MustCompile(`[^a-zA-Z0-9_/.-]+`)
```

Comment:
This regex will be compiled everytime the `sanitizeKey` is called, why not put it outside as a private global variable?